### PR TITLE
RPT_MONTH_YEAR Validation Update

### LINKED ISSUE #2793: update `dateYearIsLargerThan()` validator logic for `RPT_MONTH_YEAR`
**Description**

We need a different validator for `YYYY` in the `RPT_MONTH_YEAR` field. We currently use this:

https://github.com/raft-tech/TANF-app/blob/develop/tdrs-backend/tdpservice/parsers/validators.py#L371L376

below is an edge case:

records with `RPT_MONTH_YEAR` == `202 00` raises the following error message: `reporting month year is required but a value was not provided` and it still gets stored in the db. Couple of issues with this: 

- OFA ultimately cant do anything with these records, so these shouldnt be parsed because we cant infer the reporting period.
- the error message is unclear. a value for `RPT_MONTH_YEAR` was provided, but it is invalid (i.e. does not follow the YYYYMM format)

**Acceptance Criteria**
- [ ] records that require a `RPT_MONTH_YEAR` but do not have a valid value are not parsed. 
- [ ] error message associated with this validator is clear/user-friendly. 

**Tasks**
- [ ] Implement `preparsing_validator` with a clear error message
- [ ] Ensure the line is not parsed (no records created) and a cat 1 `ParserError` is created 
- [ ] Ensure the case aggregates counts reflect the new `row_precheck_error`

**Notes**
- relevant to T1, T2, T3, T4, T5,M1,..., M5 for TANF, SSP, and Tribal TANF
- possible approach is to implement a new `preparsing_validator` so the row is not parsed when an invalid `RPT_MONTH_YEAR` is encountered

**Supporting Documentation**

### PR DESCRIPTION
**Pull request by elipe17**
## Summary of Changes
- Added preparser validator to ensure the RPT_MONTH_YEAR field does not have any spaces.
- Added tests/updated tests
Pull request closes #2793 

## How to Test
```
cd tdrs-frontend && docker-compose up
cd tdrs-backend && docker-compose up
```

1. Open http://localhost:3000/ and sign in.
2. Submit a datafile with at least one record with an incorrect RPT_MONTH_YEAR. For example, `202 12`, or `      `.
3. Verify the record was not created
4. Verify that a pre-check parser error was created for the incorrect RPT_MONTH_YEAR
5. Verify case aggregates counts are correct

## Deliverables
_More details on how deliverables herein are assessed included [here](https://github.com/raft-tech/TANF-app/blob/develop/docs/How-We-Work/our-priorities-values-expectations.md#Deliverables)._

### [Deliverable 1: Accepted Features](https://github.com/raft-tech/TANF-app/blob/develop/docs/How-We-Work/our-priorities-values-expectations.md#Deliverable-1-Accepted-Features)

Checklist of ACs:
- [x] records that require a `RPT_MONTH_YEAR` but do not have a valid value are not parsed. 
- [x] error message associated with this validator is clear/user-friendly. 

### [Deliverable 2: Tested Code](https://github.com/raft-tech/TANF-app/blob/develop/docs/How-We-Work/our-priorities-values-expectations.md#Deliverable-2-Tested-Code)

+ Are all areas of code introduced in this PR meaningfully tested?
  + [x] If this PR introduces backend code changes, are they meaningfully tested?
  + [x] If this PR introduces frontend code changes, are they meaningfully tested?
+ Are code coverage minimums met?
  + [x] Frontend coverage: [_insert coverage %_] (see `CodeCov Report` comment in PR)
  + [x] Backend coverage: [_insert coverage %_] (see `CodeCov Report` comment in PR)

### [Deliverable 3: Properly Styled Code](https://github.com/raft-tech/TANF-app/blob/develop/docs/How-We-Work/our-priorities-values-expectations.md#Deliverable-3-Properly-Styled-Code)

+ [x] Are backend code style checks passing on CircleCI?
+ [x] Are frontend code style checks passing on CircleCI?
+ [x] Are code maintainability principles being followed?

### [Deliverable 4: Accessible](https://github.com/raft-tech/TANF-app/blob/develop/docs/How-We-Work/our-priorities-values-expectations.md#Deliverable-4-Accessibility)

+ [ ] Does this PR complete the epic? 
+ [ ] Are links included to any other gov-approved PRs associated with epic?
+ [ ] Does PR include documentation for Raft's a11y review? 
+ [ ] Did automated and manual testing with `iamjolly` and `ttran-hub` using Accessibility Insights reveal any errors introduced in this PR?

### [Deliverable 5: Deployed](https://github.com/raft-tech/TANF-app/blob/develop/docs/How-We-Work/our-priorities-values-expectations.md#Deliverable-5-Deployed)

+ [x] Was the code successfully deployed via automated CircleCI process to development on Cloud.gov?

### [Deliverable 6: Documented](https://github.com/raft-tech/TANF-app/blob/develop/docs/How-We-Work/our-priorities-values-expectations.md#Deliverable-6-Code-documentation)

+ [x] Does this PR provide background for why coding decisions were made?
+ [ ] If this PR introduces backend code, is that code easy to understand and sufficiently documented, both inline and overall?
+ [ ] If this PR introduces frontend code, is that code easy to understand and sufficiently documented, both inline and overall?
+ [ ] If this PR introduces dependencies, are their licenses documented?
+ [x] Can reviewer explain and take ownership of these elements presented in this code review?

### [Deliverable 7: Secure](https://github.com/raft-tech/TANF-app/blob/develop/docs/How-We-Work/our-priorities-values-expectations.md#Deliverable-7-Secure)

+ [x] Does the OWASP Scan pass on CircleCI?
+ [ ] Do manual code review and manual testing detect any new security issues?
+ [ ] If new issues detected, is investigation and/or remediation plan documented? 

### [Deliverable 8: User Research](https://github.com/raft-tech/TANF-app/blob/develop/docs/How-We-Work/our-priorities-values-expectations.md#Deliverable-8-User-Research)

Research product(s) clearly articulate(s):
+ [ ] the purpose of the research
+ [ ] methods used to conduct the research 
+ [ ] who participated in the research
+ [ ] what was tested and how
+ [ ] impact of research on TDP
+ [ ] (_if applicable_) final design mockups produced for TDP development

### CODE REVIEW

File: tdrs-backend/tdpservice/parsers/validators.py [hunk 1]
```diff
@@ -399,7 +399,15 @@ def validateRace():
...
-        lambda value: f"{value} is not greater than or equal to 0 or smaller than or equal to 1.",
+        lambda value: f"{value} is not greater than or equal to 0 or smaller than or equal to 2.",
+    )
+
+
+def validateRptMonthYear():
+    """Validate RPT_MONTH_YEAR."""
+    return make_validator(
+        lambda value: value[2:8].isdigit(),
+        lambda value: f"The value: {value[2:8]}, does not follow the YYYYMM format for Reporting Year and Month.",
```
Comment (andrew-jameson) [inline tdrs-backend/tdpservice/parsers/validators.py:427]: @reitermb Any input on this given the recent errors audit? I think we got flagged on saying "the value" in the past.

**Review by andrew-jameson — APPROVED** [review]
Good with me, but there is the question on the resulting error message.

File: tdrs-backend/tdpservice/parsers/test/test_parse.py [hunk 2]
```diff
@@ -389,9 +389,22 @@ def test_parse_bad_trailer_file2(bad_trailer_file_2):
...
+    trailer_error_3 = trailer_errors[3]
+    assert trailer_error_3.error_type == ParserErrorCategoryChoices.PRE_CHECK
+    assert trailer_error_3.error_message == ('The value: trash, does not follow the YYYYMM format for '
+                                             'Reporting Year and Month.')
+    assert trailer_error_3.content_type is None
+    assert trailer_error_3.object_id is None
+
+    trailer_error_4 = trailer_errors[4]
+    assert trailer_error_4.error_type == ParserErrorCategoryChoices.PRE_CHECK
+    assert trailer_error_4.error_message == 'T1trash contains blanks between positions 8 and 19.'
```
Comment (raftmsohani) [inline tdrs-backend/tdpservice/parsers/test/test_parse.py:412]: add space for error message so we have space between T1 and trash

**Review by raftmsohani — APPROVED** [review]
LGTM, except one small nit!

File: tdrs-backend/tdpservice/parsers/test/test_parse.py (hunk 2 above)
Comment (jtimpe) [reply tdrs-backend/tdpservice/parsers/test/test_parse.py:412]: i think `T1trash` is the entire/literal header record for one of our test files

File: tdrs-backend/tdpservice/parsers/validators.py [hunk 3]
```diff
@@ -399,7 +399,15 @@ def validateRace():
...
         lambda value: value >= 0 and value <= 2,
-        lambda value: f"{value} is not greater than or equal to 0 or smaller than or equal to 1.",
+        lambda value: f"{value} is not greater than or equal to 0 or smaller than or equal to 2.",
+    )
+
+
+def validateRptMonthYear():
+    """Validate RPT_MONTH_YEAR."""
+    return make_validator(
+        lambda value: value[2:8].isdigit(),
```
Comment (jtimpe) [inline tdrs-backend/tdpservice/parsers/validators.py]: is `[2:8]` always going to be the index range for `RPT_MONTH_YEAR`, in every record? if we're going to continue using this pattern for other fields, we may want to look at some sort of abstraction or modification to the `RowSchema`

**Review by jtimpe — APPROVED** [review]
one tiny question, otherwise looks good!

File: tdrs-backend/tdpservice/parsers/validators.py (hunk 3 above)
Comment (ADPennington) [reply tdrs-backend/tdpservice/parsers/validators.py]: > is `[2:8]` always going to be the index range for `RPT_MONTH_YEAR`, in every record? 

yep. for files that have this data element, its always the first data element after record type.

File: tdrs-backend/tdpservice/parsers/validators.py (hunk 1 above)
Comment (ADPennington) [reply tdrs-backend/tdpservice/parsers/validators.py:427]: @reitermb 🆙

**Review by atrimpe — APPROVED** [review]
lgtm

File: tdrs-backend/tdpservice/parsers/schema_defs/ssp/m1.py [hunk 4]
```diff
@@ -12,6 +12,7 @@
             document=SSP_M1DataSubmissionDocument(),
             preparsing_validators=[
                 validators.hasLength(150),
+                validators.validateRptMonthYear(),
```
Comment (ADPennington) [inline tdrs-backend/tdpservice/parsers/schema_defs/ssp/m1.py:15]: @elipe17 can you confirm that the logic on rpt month year field as follows:

- first check if the the field has 6 digits (any 6 digits, and not necessarily a valid YYYYMM
- if it has 6 digits, parse. the record with a potentially unusable rpt month year are stored in the db
2849p1

- if not 6 digits, do no parse the record.

**Comment by elipe17** [comment]
@reitermb we definitely need some better language for the validator now that I implemented Alex's requests. The validator checks that there are 6 digits in the field, that the year in the field is valid and that the month in the field is valid. Below are some examples of the failed output with respect to the year being incorrect and the month being incorrect.  cc @ADPennington 

```
The value: 189910, does not follow the YYYYMM format for Reporting Year and Month.

The value: 202015, does not follow the YYYYMM format for Reporting Year and Month.
```

**Comment by ADPennington** [comment]
> @reitermb we definitely need some better language for the validator now that I implemented Alex's requests. The validator checks that there are 6 digits in the field, that the year in the field is valid and that the month in the field is valid. Below are some examples of the failed output with respect to the year being incorrect and the month being incorrect. cc @ADPennington
> 
> ```
> The value: 189910, does not follow the YYYYMM format for Reporting Year and Month.
> 
> The value: 202015, does not follow the YYYYMM format for Reporting Year and Month.
> ```

@reitermb also worth noting that this will appear like the following in the error report. let us know if this is okay. 

2849p2

**Comment by reitermb** [comment]
> > @reitermb we definitely need some better language for the validator now that I implemented Alex's requests. The validator checks that there are 6 digits in the field, that the year in the field is valid and that the month in the field is valid. Below are some examples of the failed output with respect to the year being incorrect and the month being incorrect. cc @ADPennington
> > ```
> > The value: 189910, does not follow the YYYYMM format for Reporting Year and Month.
> > 
> > The value: 202015, does not follow the YYYYMM format for Reporting Year and Month.
> > ```
> 
> @reitermb also worth noting that this will appear like the following in the error report. let us know if this is okay.
> 
> 2849p2

@elipe17 @ADPennington 

So what this is communicating is that the year is out of range, right? Meaning either:
1. The year doesn't match the chosen year in TDP
2. The month doesn't exist, e.g. "13" or "00"

But that there _are_ 4 digits in the YYYY range and two digits in the MM range. 

How hard would it be to add on valid options to that existing error since we know (from TDP data file selection) what reporting period bucket people should be trying to upload data for. e.g.,

`The value: [string], does not follow a valid YYYYMM format for Reporting Year and Month. Valid values for [Plain language Quarter, Year] are [string1], [string2], [string3]` 

translating to something like:
`The value: 202015 does not follow the YYYYMM format for Reporting Year and Month. Valid values for Fiscal Quarter 1, 2020 are 202010, 202011, and 202012`

**Comment by elipe17** [comment]
> > > @reitermb we definitely need some better language for the validator now that I implemented Alex's requests. The validator checks that there are 6 digits in the field, that the year in the field is valid and that the month in the field is valid. Below are some examples of the failed output with respect to the year being incorrect and the month being incorrect. cc @ADPennington
> > > ```
> > > The value: 189910, does not follow the YYYYMM format for Reporting Year and Month.
> > > 
> > > The value: 202015, does not follow the YYYYMM format for Reporting Year and Month.
> > > ```
> > 
> > 
> > @reitermb also worth noting that this will appear like the following in the error report. let us know if this is okay.
> > 2849p2
> 
> @elipe17 @ADPennington
> 
> So what this is communicating is that the year is out of range, right? Meaning either:
> 
> 1. The year doesn't match the chosen year in TDP
> 2. The month doesn't exist, e.g. "13" or "00"
> 
> But that there _are_ 4 digits in the YYYY range and two digits in the MM range.
> 
> How hard would it be to add on valid options to that existing error since we know (from TDP data file selection) what reporting period bucket people should be trying to upload data for. e.g.,
> 
> `The value: [string], does not follow a valid YYYYMM format for Reporting Year and Month. Valid values for [Plain language Quarter, Year] are [string1], [string2], [string3]`
> 
> translating to something like: `The value: 202015 does not follow the YYYYMM format for Reporting Year and Month. Valid values for Fiscal Quarter 1, 2020 are 202010, 202011, and 202012`

@reitermb, That is possible but would duplicate a lot/all of the work done in [this PR](https://github.com/raft-tech/TANF-app/pull/2838). Once 2838 merges we could update this PR to take those field values and impute them into the validator. I'd like to see what @ADPennington thinks about that.

**Comment by ADPennington** [comment]
@elipe17 @raftmsohani getting an elastic-related failure in the dev deployment pipeline.

**Comment by ADPennington** [comment]
> > > > @reitermb we definitely need some better language for the validator now that I implemented Alex's requests. The validator checks that there are 6 digits in the field, that the year in the field is valid and that the month in the field is valid. Below are some examples of the failed output with respect to the year being incorrect and the month being incorrect. cc @ADPennington
> > > > ```
> > > > The value: 189910, does not follow the YYYYMM format for Reporting Year and Month.
> > > > 
> > > > The value: 202015, does not follow the YYYYMM format for Reporting Year and Month.
> > > > ```
> > > 
> > > 
> > > @reitermb also worth noting that this will appear like the following in the error report. let us know if this is okay.
> > > 2849p2
> > 
> > 
> > @elipe17 @ADPennington
> > So what this is communicating is that the year is out of range, right? Meaning either:
> > 
> > 1. The year doesn't match the chosen year in TDP
> > 2. The month doesn't exist, e.g. "13" or "00"
> > 
> > But that there _are_ 4 digits in the YYYY range and two digits in the MM range.
> > How hard would it be to add on valid options to that existing error since we know (from TDP data file selection) what reporting period bucket people should be trying to upload data for. e.g.,
> > `The value: [string], does not follow a valid YYYYMM format for Reporting Year and Month. Valid values for [Plain language Quarter, Year] are [string1], [string2], [string3]`
> > translating to something like: `The value: 202015 does not follow the YYYYMM format for Reporting Year and Month. Valid values for Fiscal Quarter 1, 2020 are 202010, 202011, and 202012`
> 
> @reitermb, That is possible but would duplicate a lot/all of the work done in [this PR](https://github.com/raft-tech/TANF-app/pull/2838). Once 2838 merges we could update this PR to take those field values and impute them into the validator. I'd like to see what @ADPennington thinks about that.

@elipe17 @reitermb please add this as a task in the cat 1 cleanup ticket #2838

**Review by ADPennington — APPROVED** [review]
this looks good to me @elipe17 🚀 

the cleanup of the error message can be handled in #2838; else we'll need another ticket. 

testing notes [here](https://hackmd.io/P4sQkTRORLGjiodfgRwMtA?view#RPT_MONTH_YEAR-Validation-Update)

**Pull request merged**